### PR TITLE
.

### PR DESCRIPTION
**Pull request by eking1082**

### CODE REVIEW

**Comment by ziplokk1** [comment]
Just curious, why did you do everything in one commit?

**Comment by ziplokk1** [comment]
What is your reasoning for using mongodb?

File: README.md [hunk 1]
```diff
@@ -1,78 +1,45 @@
...
-As per our [culture](https://www.cyberfortress.com/about/culture/), 
-we base all merit upon what you are capable of and not upon
-any certifications or education. Being counter-culture, consider this your
-resume. Given that, we thought that this would 
-be a creative way to express that and allow you, the applicant, 
-to express yourself. 
+* Node v12 is assumed, you can use [NVM](https://github.com/nvm-sh/nvm#install--update-script) to install if needed
+* Ensure that the [MongoDB](https://www.mongodb.com/download-center/community) community server is installed
+* Run `npm install` and `npm run db:seed` from the app directory
+* Create the env file with `cp .env.example .env`, I will provide my Geocodio and Darksky keys via email
```
Comment (ziplokk1) [inline README.md:9]: lines 7 & 8 should be swapped. I was having trouble getting it running without the `.env` file present.

File: README.md [hunk 2]
```diff
@@ -1,78 +1,45 @@
...
-    generate your own user dataset. (Please scrape responsibly. 
-    Don't scrape too much or too often. I've emailed the dev to let me buy
-    him some beers in preparation for this, but he hasn't got back to me yet
-    and I'd like to not spend hundreds of dollars on beers for our friend.)
-
-## Credits
-
-Addresses, names, and emails were generated using [randomlists.com](https://www.randomlists.com/random-addresses)
\ No newline at end of file
+Use `npm test` to run all of the tests and show coverage.
```
Comment (ziplokk1) [inline README.md:45]: This requires the `.env` file to be present as well or the tests will not run.

File: config/passport.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  usernameField: 'email',
+}, (email, password, done) => {
+  Admin.findOne({ email }, (err, admin) => {
+    if (err) return done(err);
+    if (!admin) return done(null, false, { msg: `Email ${email} not found.` });
+
+    admin.comparePassword(password, (err, isMatch) => {
+      if (err) { return done(err); }
+      if (isMatch) return done(null, admin);
+      return done(null, false, 'Invalid email or password.');
```
Comment (ziplokk1) [inline config/passport.js:20]: This line is untested. In what case would this line ever be called?

**Comment by eking1082** [comment]
> Just curious, why did you do everything in one commit?

In my experience the convention for pull requests is to squash your commits into one.

> What is your reasoning for using mongodb?

I chose mongodb because I typically use Postgres and I wanted to learn more about mongo. I feel that either a SQL or NoSQL database is suited for the problem, so I took the opportunity to learn something new.

File: README.md [hunk 4]
```diff
@@ -1,78 +1,45 @@
...
+```
+curl -X POST 'http://localhost:8080/admin/signup?email=test@test.com&password=password'
+```
 
-Your mission, should you choose to accept it, is to build an API which, 
-given a user's unique identifier, should tell us whether it is raining
-in the location where the user lives.
+###### Login with your new admin user
+```
+curl -X POST 'http://localhost:8080/admin/login?email=test@test.com&password=password'
```
Comment (ziplokk1) [inline README.md:21]: Why is the request method a POST with no request data?

File: README.md (hunk 1 above)
Comment (eking1082) [reply README.md]: Ah yes, that's correct. My mistake.

File: README.md (hunk 2 above)
Comment (eking1082) [reply README.md]: Correct, the env file is required to run the tests.

File: config/passport.js (hunk 3 above)
Comment (eking1082) [reply config/passport.js:20]: I did not write a test for that scenario. That is reached if the login request has an invalid password. I have committed a test for that scenario in my repo.

File: README.md (hunk 4 above)
Comment (eking1082) [reply README.md]: The request data is in the query string, I suppose technically that should be in the body for a POST.

File: config/passport.js (hunk 3 above)
Comment (ziplokk1) [reply config/passport.js:20]: You didn't have to fix the test. I was just curious if there was any reason that it was omitted.

**Comment by ziplokk1** [comment]
What is the branch coverage percentage for your PR?

File: app.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+const adminController = require('./controllers/admin');
+const userController = require('./controllers/user');
+
+const app = express();
+
+app.set('port', process.env.PORT || 8080);
+
+app.use(logger('dev'));
+app.use(bodyParser.json());
+app.use(express.json());
```
Comment (ziplokk1) [inline app.js]: Why use both `bodyParser.json()` and `express.json()` middleware?

**Comment by eking1082** [comment]
> What is the branch coverage percentage for your PR?

The branch coverage is ~69%. The line coverage is ~92%.

File: app.js (hunk 5 above)
Comment (eking1082) [reply app.js]: Good catch, `app.use(express.json());` is not necessary. I included body-parser later in the development process and missed that I was using both middlewares.

**Comment by ziplokk1** [comment]
> The branch coverage is ~69%. The line coverage is ~92%.

One of the requirements of the project is to have at least 80%. Could you update your PR to fix that?

**Comment by eking1082** [comment]
I have written additional tests and coverage is now 100%. Without mocking packages I am unable to produce tests for some of the package level exceptions so I have used Istanbul sparingly where necessary.

**Comment by ziplokk1** [comment]
Awesome, thanks man! I'll reach out to you via email from here out.

**Pull request closed without merging**